crates/sui-indexer-alt-framework/src/store.rs

```diff
@@ -0,0 +1,173 @@
+    /// the pruner last updated the bounds, and the configured `delay`. More specifically, this is
+    /// the result of delay + (pruner_timestamp - current_database_time)
+    ///
+    /// The pruner is allowed to prune the region between the returned `pruner_hi` (inclusive) and
+    /// `reader_lo` (exclusive) after `wait_for` milliseconds have passed since this response was
+    /// returned.
+    async fn pruner_watermark(
+        &mut self,
+        pipeline: &'static str,
+        delay: Duration,
+    ) -> anyhow::Result<Option<PrunerWatermark>>; // (pruner_hi, reader_lo, wait_for_ms)
+
+    /// Update the pruner watermark, returns true if the watermark was actually updated
+    async fn set_pruner_watermark(
+        &mut self,
+        pipeline: &'static str,
+        pruner_hi: i64,
+    ) -> anyhow::Result<bool>;
+}
+
+/// A storage-agnostic interface for managing pipeline watermarks.
+///
+/// This trait abstracts away the underlying storage implementation, providing
+/// a consistent way to connect to the database regardless of the specific
+/// storage technology being used.
+#[async_trait]
+pub trait Store: Send + Sync + 'static + Clone {
+    type Connection<'c>: DbConnection
+    where
+        Self: 'c;
+
+    async fn connect<'c>(&'c self) -> Result<Self::Connection<'c>, anyhow::Error>;
+}
+
+pub type HandlerBatch<H> = <H as SequentialHandler>::Batch;
+
+/// Extends the Store trait with transactional capabilities.
+///
+/// This trait provides methods to execute operations within a database transaction,
+/// ensuring atomicity when committing handler batches and updating watermarks.
+/// It allows for safely combining multiple database operations that must succeed
+/// or fail together.
+#[async_trait]
+pub trait TransactionalStore: Store {
+    async fn transaction<'a, R, F>(&self, f: F) -> anyhow::Result<R>
+    where
+        R: Send + 'a,
+        F: Send + 'a,
+        F: for<'r> FnOnce(
+            &'r mut Self::Connection<'_>,
+        ) -> ScopedBoxFuture<'a, 'r, anyhow::Result<R>>;
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+pub struct CommitterWatermark {
+    pub epoch_hi_inclusive: i64,
+    pub checkpoint_hi_inclusive: i64,
+    pub tx_hi: i64,
+    pub timestamp_ms_hi_inclusive: i64,
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+pub struct ReaderWatermark {
+    /// Within the framework, this value is used to determine the new `reader_lo`.
+    pub checkpoint_hi_inclusive: i64,
+    /// Within the framework, this value is used to check whether to actually make an update
+    /// transaction to the database.
+    pub reader_lo: i64,
+}
+
+#[derive(Default, Debug, Clone, Copy)]
+pub struct PrunerWatermark {
+    /// How long to wait from when this query ran on the database until this information can be
+    /// used to prune the database. This number could be negative, meaning no waiting is necessary.
+    pub wait_for: i64,
```

Comment:
```suggestion
    pub wait_for_ms: i64,
```

(and mention the unit in the doc comment).